src/parse_async.cc

```diff
@@ -12,114 +12,170 @@ using Nan::AsyncWorker;
+
+                if (0 != parse_err_code) {
+                    argv[0] = annotations2v8(result);
+                }
+
+                callback->Call(2, argv);
+                return;
+            }
+            Nan::ThrowTypeError("not handled OKCallback");
         }
 
-        if (sourceData) {
-            delete sourceData;
-            sourceData = nullptr;
+        virtual ~ParseWorker()
+        {
+            if (result) {
+                drafter_free_result(result);
+                result = nullptr;
+            }
+
+            if (sourceData) {
+                delete sourceData;
+                sourceData = nullptr;
+            }
         }
-    }
-};
+    };
+}
+
+/**
+ * posible args variant
+ * (source) -> promise
+ * (source, options) -> promise
+ * (source, callback) -> callback
+ * (source, option, callback) -> callback
+ */
+
+static bool isLastParamCallback(const Nan::FunctionCallbackInfo<v8::Value>& info)
+{
+    return info[info.Length()-1]->IsFunction();
 }
 
-NAN_METHOD(protagonist::Parse) {
+NAN_METHOD(protagonist::Parse)
+{
     Nan::HandleScope scope;
+    Nan::Persistent<v8::Promise::Resolver>* persistent = nullptr;
+    Callback* callback = nullptr;
+    v8::Local<v8::Promise::Resolver> resolver;
 
     // Check arguments
-    if (info.Length() != 2 && info.Length() != 3) {
-        Nan::ThrowTypeError(
-            "wrong number of arguments, `parse(string, options, callback)` "
-            "expected");
+
+    size_t optionIndex = 0;
+    size_t callbackIndex = 0;
+
+    if (info.Length() < 1 || info.Length() > 3) {
+        Nan::ThrowTypeError("wrong number of arguments, `parse(string, options, callback)` expected");
         return;
     }
 
     if (!info[0]->IsString()) {
-        Nan::ThrowTypeError(
-            "wrong argument - string expected, `parse(string, options, "
-            "callback)`");
+        Nan::ThrowTypeError("wrong 1st argument - string expected");
         return;
     }
 
-    if ((info.Length() == 2 && !info[1]->IsFunction()) ||
-        (info.Length() == 3 && !info[2]->IsFunction())) {
-        Nan::ThrowTypeError(
-            "wrong argument - callback expected, `parse(string, options, "
-            "callback)`");
-        return;
+    if (isLastParamCallback(info)) { // last arg is callback funtion
+        if (info.Length() == 3) {
+            optionIndex = 1;
+        }
+        callbackIndex = info.Length() - 1;
+    } else { // is Promise
+        if (info.Length() == 2) {
+            optionIndex = 1;
+        } else if (info.Length() > 2) { // promise shoud not have more than 2 params
+            Nan::ThrowSyntaxError("wrong number of arguments, `parse(string, [options])` expected");
+            return;
+        }
     }
 
-    if (info.Length() == 3 && !info[1]->IsObject()) {
-        Nan::ThrowTypeError(
-            "wrong argument - object expected, `parse(string, options, "
-            "callback)`");
+    if (optionIndex && !info[optionIndex]->IsObject()) {
```

Comment:
You don't need the `optionIndex`. It will always be `1`. You just need to know if it exists. I think that is a cleaner way of logic. So, rename it `optionsExist` of boolean type